### PR TITLE
Improve IO.forkAll so it returns a composite fiber

### LINKED ISSUE #18: IO.forkAll should return a composite fiber
Right now we have:

```scala
def forkAll[E, A](ios: List[IO[E, Unit]]): IO[E, Unit]
```

Instead we should have:

```scala
def forkAll[E, E2, A](as: List[IO[E, A]]): IO[E2, Fiber[E, List[A]]]
```

### PR DESCRIPTION
**Pull request by DennisVDB**
Here's my implementation of `IO.forkAll` that returns a composite fiber (closes #18).
I used `foldRight` so the accumulated values are in the right order (`::` is right associative).

### CODE REVIEW

**Comment by jdegoes** [comment]
Can you please resolve conflicts? Thanks!

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 1]
```diff
@@ -775,10 +775,14 @@ object IO {
...
-  def forkAll[E2](l: List[IO[E2, Unit]]): IO[E2, Unit] = l match {
-    case Nil     => IO.unit[E2]
-    case x :: xs => x.fork *> forkAll(xs)
-  }
+  final def forkAll[E, E2, A](as: List[IO[E, A]]): IO[E2, Fiber[E, List[A]]] =
+    as.foldRight(IO.point[E2, Fiber[E, List[A]]](Fiber.point(List.empty))) {
+      case (a, fiberAs) =>
+        for {
+          fiberA   <- a.fork
+          fiberAs0 <- fiberAs.map(_.zipWith(fiberA)((as, a) => a :: as))
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: Can you rewrite this using flatMap so it becomes tail recursive?

File: core/shared/src/main/scala/scalaz/zio/IOQueue.scala [hunk 2]
```diff
@@ -85,7 +85,7 @@ class IOQueue[A] private (capacity: Int, ref: IORef[State[A]]) {
   final def interruptTake[E](t: Throwable): IO[E, Boolean] =
     IO.flatten(ref.modifyFold[E, IO[E, Boolean]] {
       case Deficit(takers) if takers.nonEmpty =>
-        val forked: IO[E, Unit] = IO.forkAll(takers.toList.map(_.interrupt[E](t).toUnit))
+        val forked: IO[E, Unit] = IO.forkAll(takers.toList.map(_.interrupt[E](t))).toUnit
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/IOQueue.scala]: In this and the other change, place a TODO comment to use the composite Fiber. Doesn’t have to be done now but will be useful.

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 1 above)
Comment (DennisVDB) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: I need to rewrite the for-comprehension so `zipWith` is the last call in the for comprehension or the entire `foldRight`?

Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: In this case, I just mean without the final useless ‘map’ that Scala’s desugaring will insert.

Comment (hrhino) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: We should consider adding oleg-py/better-monadic-for for stuff like that.

**Comment by jdegoes** [comment]
@dennisVDB Thanks for the contribution! 😄

**Comment by ktonga** [comment]
Why are we specializing on `List` here while everywhere else `TraversableOnce` is accepted?

**Comment by DennisVDB** [comment]
That’s true, I kind of just followed the definition given in #18. Should I make a mini-PR that fixes that? 

It was a nice experience for a first open source PR!

**Comment by hrhino** [comment]
> Why are we specializing on `List` here while everywhere else `TraversableOnce` is accepted?

Why are we accepting `TraversableOnce` everywhere else?

**Comment by ktonga** [comment]
@hrhino rule of least power I guess

**Pull request merged**